src/runtime/database.cpp

```diff
@@ -1407,49 +1319,142 @@ static std::vector<FileDependency> get_all_file_dependencies(const Database *db,
+  // --------------------------------------------------------------------------------------------------------------------------------
+  // |    1   |  foo  |   1    |    1   |  1234   | ls lah .    |   0    | <d>a.txt<d>b.txt<d> |
+  // <d>c.o<d>d.o<d> | <d>a=b<d>c=d<d>  | |    1   |  foo  |   1    |    1   |  1234   | ls lah . |
+  // 0    | <d>a.txt<d>b.txt<d> | <d>c.o<d>d.o<d> | <d>a=b<d>c=d<d>  |
+  //
+  // The subtable is constructed by joining the jobs table with the minimal set of other dependent
+  // tables with the following extra processing.
+  // 1. files table is split into 2 columns, input_files & output_files
+  // 2. tags are flattened from two columns (uri, content) to one column (tags) with a = separator
+  // 3. input_files, output_files, and tags are group_concat'd into a single row per job. <d> is
+  //    used as a deliminator between each value. The deliminator is also places at the beginning
+  //    and end of each row so that queries don't need to special case the first/last entry.
+  //
+  // Any inspection flag/user code may add any WHERE expression conditions to the main query using
+  // the columns of the subtable for fine grain filters.
+  //
+  // For example, the query below will return all jobs that exited with status code 0 and output at
+  // least one .0 file.
+  //   SELECT job_id FROM **SUBTABLE**
+  //   WHERE status = 0 AND label like '%<d>%.o<d>%'
+  std::string subtable = R"delim(
+          SELECT 
+              j.job_id, 
+              j.label, 
+              j.run_id, 
+              j.use_id, 
+              j.endtime, 
+              j.commandline, 
+              s.status, 
+              '<d>' || group_concat(ft_input.path, '<d>') || '<d>' input_files, 
+              '<d>' || group_concat(ft_output.path,'<d>') || '<d>' output_files, 
+              '<d>' || group_concat(t.tag, '<d>') || '<d>' tags
+          FROM jobs j
+          LEFT JOIN (
+              SELECT stat_id, status FROM stats
+          ) s
+          ON j.stat_id=s.stat_id
+          LEFT JOIN (
+              SELECT job_id, uri || '=' || content tag FROM tags
+          ) t
+          ON j.job_id = t.job_id
+          LEFT JOIN (
+              SELECT filetree.job_id, files.path FROM filetree
+              INNER JOIN files
+              ON filetree.file_id=files.file_id
+              WHERE access = 1
+          ) ft_input
+          ON j.job_id=ft_input.job_id
+          LEFT JOIN (
+              SELECT filetree.job_id, files.path FROM filetree
+              INNER JOIN files
+              ON filetree.file_id=files.file_id
+              WHERE access = 2
+          ) ft_output
+          ON j.job_id=ft_output.job_id
+          GROUP BY
+              j.job_id
+)delim";
+
+  // This query wraps the subtable, applies the requested filters, and returns the matching jobs
+  std::string id_query = "SELECT job_id\nFROM (\n" + subtable + ")\nWHERE\n\t";
+  id_query += collapse_and(and_or_filters);
```

Comment:
If `collapse_and` returns the empty string does `WHERE` still end the id_query correctly? Or do we need a guard on this `WHERE`s inclusion? Or maybe there is no way to end up here with zero filters?

Reply:
It's guarded as you saw in a comment below